cmds/dhclient/dhcp6client/packet.go

```diff
@@ -0,0 +1,179 @@
+		Version:      ipv6.Version,
+		TrafficClass: int((hb[0]&0x0f)<<4 | (hb[1]&0xf0)>>4),
+		FlowLabel:    int(flowlbl),
+		PayloadLen:   int(hb[4]<<8 | hb[5]),
+		NextHeader:   int(hb[6]),
+		HopLimit:     int(hb[7]),
+		Src:          hb[8:24],
+		Dst:          hb[24:40],
+	}
+}
+
+/***
+* UDP header
+ */
+func marshapUdpHdr(h *Udphdr) []byte {
+	udphdr := make([]byte, udpHdrLen)
+
+	// src port
+	binary.BigEndian.PutUint16(udphdr[0:2], h.Src)
+	// dest port
+	binary.BigEndian.PutUint16(udphdr[2:4], h.Dst)
+	// length
+	binary.BigEndian.PutUint16(udphdr[4:6], h.Length)
+	return udphdr
+}
+
+func unmarshalUdpHdr(hb []byte) *Udphdr {
+	return &Udphdr{
+		Src:    binary.BigEndian.Uint16(hb[0:2]),
+		Dst:    binary.BigEndian.Uint16(hb[2:4]),
+		Length: binary.BigEndian.Uint16(hb[4:6]),
+		Csum:   binary.BigEndian.Uint16(hb[6:8]),
+	}
+}
+
+/***
+* DHCP packet
+ */
+func addSolicitOptions(mac *net.HardwareAddr) (dhcp6.Options, error) {
+	// make options: iata
+	var id = [4]byte{0x00, 0x00, 0x00, 0x0f}
+	options := make(dhcp6.Options)
+	if err := options.Add(dhcp6.OptionIANA, dhcp6.NewIANA(id, 0, 0, nil)); err != nil {
+		return nil, err
+	}
+	// make options: rapid commit
+	if err := options.Add(dhcp6.OptionRapidCommit, nil); err != nil {
+		return nil, err
+	}
+	// make options: elapsed time
+	var et dhcp6.ElapsedTime
+	et.UnmarshalBinary([]byte{0x00, 0x00})
+	if err := options.Add(dhcp6.OptionElapsedTime, et); err != nil {
+		return nil, err
+	}
+	// make options: option request option
+	oro := make(dhcp6.OptionRequestOption, 4)
+	oro.UnmarshalBinary([]byte{0x00, 0x17, 0x00, 0x18})
+	if err := options.Add(dhcp6.OptionORO, oro); err != nil {
+		return nil, err
+	}
+	// make options: duid with mac address
+	duid := dhcp6.NewDUIDLL(6, *mac)
+	db, err := duid.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	// add row
+	options[dhcp6.OptionClientID] = append(options[dhcp6.OptionClientID], db)
+
+	return options, nil
+}
+
+func newDhcpPacket(messageType dhcp6.MessageType, txID [3]byte, addr *net.UDPAddr, options dhcp6.Options) *dhcp6.Packet {
+	return &dhcp6.Packet{
+		MessageType:   messageType,
+		TransactionID: txID,
+		Options:       options,
+	}
+}
+
+/***
+* Helper functions
+ */
+func doCsum(ipv6hdr []byte, udphdr []byte, packet []byte) []byte {
+	csum := make([]byte, 2)
+	// psuedoheader = srcip + dstip + udpcode + udplen
+	psuedoHdr := append(ipv6hdr[8:], []byte{0x00, 0x11 /*udpcode=17*/}...)
+	psuedoHdr = append(psuedoHdr, udphdr[4:6]...)
+	// udp header + data (excluding checksum)
+	udpData := append(udphdr, packet...)
+	// calculate 16-bit sum
+	sumPsuedoHdr := sixteenBitSum(psuedoHdr)
+	sumUdpData := sixteenBitSum(udpData)
+	sumTotal := sumPsuedoHdr + sumUdpData
+	sumTotal = (sumTotal>>16 + sumTotal&0xffff)
+	sumTotal = sumTotal + (sumTotal >> 16)
+	// one's complement
+	sumTotal = ^sumTotal
+	csum[0] = uint8(sumTotal & 0xff)
+	csum[1] = uint8(sumTotal >> 8)
+	return csum
+}
+
+func sixteenBitSum(p []byte) uint32 {
```

Comment:
I did not find it when I was implementing it. Since it works, can we leave it here for now?